### PR TITLE
Use payload, not generate, in payload modules

### PR DESCRIPTION
**Pull request by scriptjunkie**
This effectively makes any changes to shellcode before passing to upstream
classes instead of after. PrependMigrate had been failing because it was
miscalculating payload sizes. This will fix that issue.

### CODE REVIEW

**Comment by bcook-r7** [comment]
Is it fair to say, in order to verify that this doesn't break anything, we need a test that generates all of the payload/stagers with deterministic inputs (e.g. seed the random number generator) and compare the outputs before and after this test?

**Comment by scriptjunkie** [comment]
No, this only changes the code to 9 payloads:
windows/exec
windows/loadlibrary
windows/x64/exec
windows/x64/loadlibrary
windows/reverse_http
windows/reverse_https
windows/reverse_https_proxy
windows/reverse_tcp_dns
windows/x64/reverse_https

So you can generate all the affected payloads/stagers with this command:
```
setg CMD calc
setg DLL test.dll
setg LHOST 192.168.12.34

use payload/windows/exec
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
use payload/windows/loadlibrary
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
use payload/windows/x64/exec
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
use payload/windows/x64/loadlibrary
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
use payload/windows/meterpreter/reverse_http
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
use payload/windows/meterpreter/reverse_https
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
use payload/windows/meterpreter/reverse_https_proxy
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
use payload/windows/meterpreter/reverse_tcp_dns
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
use payload/windows/x64/meterpreter/reverse_https
generate -t raw -f /tmp/msf.raw
md5sum /tmp/msf.raw >> /tmp/before.txt
```
Check out the other repository, and repeat. You'll find the only MD5's that differ are the ones that are the *reverse_http* ones. Next generate those, comparing the msf.raw files that are generated and you'll find the only bytes that differ are the 4 randomly generated initial request URL bytes, which correspond to the generate_uri_checksum() method in the payloads.

So bottom line, yes, you can and I have generate all of the payloads changed by this commit and yes, it does not affect them. The only change that happens is if you use prepend_migrate. In that case, it will be broken without this fix and not be broken with this fix. This fix does not break anything.

**Comment by jlee-r7** [comment]
This abuses the `payload` method, intended to be a static blob of shellcode, which fundamentally alters the way compatibility is determined. Seeing matching hashes for before and after will go a long way to assuage my fears. What I would also like to see is that the payload cache does not hork these up, i.e., with a different `CMD` do we get different exec payloads?

Also, it affects payloads that `include` the changed mixins, not just the payloads whose code was directly changed. After a quick grep, the only one i see that you missed is modules/payloads/singles/windows/adduser.rb

**Comment by scriptjunkie** [comment]
The hashes are as follows (I ran again with adduser included):
before (generated from git master branch):
```
bd34e21bb7acb0e565fc228752de6579  payload/windows/exec
e0949fae973e14f29a189fc05f2d9dfa  payload/windows/loadlibrary
527c1e4aff548b599002bb93f5a166a1  payload/windows/x64/exec
4de59d818abc5394c3d4c1e59dd4fb31  payload/windows/x64/loadlibrary
6bb7953512a9df86f67646ab419f5910  payload/windows/meterpreter/reverse_tcp_dns
c13df4f0828fb106cc5aeeba434d7782  payload/windows/adduser

b599ba0568c91cca4e6ed284b7d7aa06  payload/windows/meterpreter/reverse_http
c0d6e532b758d1f0c2aa3dd856ad4273  payload/windows/meterpreter/reverse_https
ccee005268521b15a33a6631d3373ddb  payload/windows/meterpreter/reverse_https_proxy
6ae540bbcc0cda63387db475384ee591  payload/windows/x64/meterpreter/reverse_https
```

and after (from this branch):
```
bd34e21bb7acb0e565fc228752de6579  payload/windows/exec
e0949fae973e14f29a189fc05f2d9dfa  payload/windows/loadlibrary
527c1e4aff548b599002bb93f5a166a1  payload/windows/x64/exec
4de59d818abc5394c3d4c1e59dd4fb31  payload/windows/x64/loadlibrary
6bb7953512a9df86f67646ab419f5910  payload/windows/meterpreter/reverse_tcp_dns
c13df4f0828fb106cc5aeeba434d7782  payload/windows/adduser

30d8460087b5ab3f0f0c1e6850deee2b  payload/windows/meterpreter/reverse_http
fdd4de5ba038b4ffd77fec59d9fe2628  payload/windows/meterpreter/reverse_https
5e5bb8aef7e3b0c9dd9a8df67daa6ae0  payload/windows/meterpreter/reverse_https_proxy
c26549068ea37497cf1aa3980039d2b2  payload/windows/x64/meterpreter/reverse_https
```

The only payloads with a difference are the reverse http ones. To see the binary difference, I ran each file through XXD and then did a diff. For example:
```
$ xxd beforeX64reverseHTTPS.raw >beforeX64reverseHTTPS.xxd
$ xxd afterX64reverseHTTPS.raw >afterX64reverseHTTPS.xxd
$ diff beforeX64reverseHTTPS.xxd afterX64reverseHTTPS.xxd 
29c29
< 00001c0: 2f67 5768 3600 0049 bef0 b5a2 5600 0000  /gWh6..I....V...
---
> 00001c0: 2f50 6344 6500 0049 bef0 b5a2 5600 0000  /PcDe..I....V...
```

If you compare the XXD against the module source, you'll see that the "/12345" (or \x2f\x31\x32\x33\x34\x35) has been replaced:
XXD:
```
00001a0: ba2d 0618 7b00 0000 00ff d585 c075 2448  .-..{........u$H
00001b0: ffcf 7413 ebb1 e981 0000 00e8 7fff ffff  ..t.............
00001c0: 2f67 5768 3600 0049 bef0 b5a2 5600 0000  /gWh6..I....V...
00001d0: 00ff d548 31c9 48ba 0000 4000 0000 0000  ...H1.H...@.....
```
Module source:
```
"\xE9\x81\x00\x00\x00"+
"\xE8\x7F\xFF\xFF\xFF" + #updated jump offset
"\x2F\x31\x32\x33\x34\x35\x00" +
"\x49\xBE\xF0\xB5\xA2\x56\x00\x00\x00\x00\xFF\xD5\x48\x31\xC9\x48" +
```

So basically everything that we want to stay the same stays the same, and the only change is the expected change.

And finally here is the different-parameter test with the patch applied, showing cache isn't dorked:
```
msf > use payload/windows/adduser 
msf payload(adduser) > set USER egyp7
USER => egyp7
msf payload(adduser) > generate -t raw -f /tmp/adduserEgyp7.raw
[*] Writing 272 bytes to /tmp/adduserEgyp7.raw...
msf payload(adduser) > set USER scriptjunkie
USER => scriptjunkie
msf payload(adduser) > generate -t raw -f /tmp/adduserScript.raw
[*] Writing 286 bytes to /tmp/adduserScript.raw...
msf payload(adduser) > use payload/windows/exec
msf payload(exec) > set CMD calc
CMD => calc
msf payload(exec) > generate -t raw -f /tmp/cmdCalc.raw
[*] Writing 189 bytes to /tmp/cmdCalc.raw...
msf payload(exec) > set CMD notepad
CMD => notepad
msf payload(exec) > generate -t raw -f /tmp/cmdNote.raw
[*] Writing 192 bytes to /tmp/cmdNote.raw...
```

**Comment by todb-r7** [comment]
So...

@scriptjunkie has this change that he says -- and looks like proves -- is pretty harmless, and fixes a bug. @jlee-r7 says that this might break something else. @hmoore-r7 was also concerned about the inherent fragility of things like PrependMigrate, and @Meatballs1 says that PrependMigrate is already broken anyway.

Is there a test that can prove that:

  - [ ] This change is needed (I think it's documented here, actually)
  - [ ] This change does break something else?

Everyone on this this PR is super smart but it looks like there's a divide on opinion here. How do we throw in some facts?

**Comment by scriptjunkie** [comment]
-I'm not sure what else I can do to show it a. fixes bugs and b. doesn't break anything aside from maybe stepping through source:
    -Generate in Msf::Payload::Windows calls generate in Msf::Payload which calls internal_generate which calls payload and then returns a duplicated buffer. There just isn't much logic that differs between before and after the patch; only a few different lines are executed. But it makes payload generation more consistent; these same lines will get executed for both static blob payloads and the other payloads, instead of just for one of them. It's eliminating the inconsistencies.

-These bugs have been shown to affect, and even been reported by, users.

-The most repeated criticism that I have heard seems to be pointed at PrependMigrate, (i.e. payload size significantly differs depending on whether that option was selected) but as this pull neither implements nor removes PrependMigrate, just fixes a bug, I don't think it's fair to block it based on that.

**Comment by hmoore-r7** [comment]
Will review this once more and merge if it looks good. Thank you for your patience!

**Comment by wvu-r7** [comment]
Thanks, @hmoore-r7!

**Comment by hmoore-r7** [comment]
I am still scratching my head over why you would want to rename generate() to payload() to fix this issue. It seems like there is a better solution that doesn't change how we return code from our payload modules. Still digging, sorry.

**Comment by hmoore-r7** [comment]
OK, I think the grok the issue:
* Most payload modules that call generate() call super() first to get the "base"
* The implementation of prepends() is done by hooking generate() in an included mixin
* The PrependMigrate code broken because super() is being called before the payload is even generated.

More importantly, any module that doesn't bother calling super() will not get Prepend* support at all. The recent refactor of reverse_http(s) removed the otherwise useless call to super() since all offset patching was replaced by Metasm. Those modules now longer support PrependMigrate (although they didn't really support it before, due to this bug).

We have ~3 general classes of payloads today:
 * Those without generate() which depend on parent classes to substitute offsets into blobs for them.
 * Those which implement generate() but first call super() to get any upstream generated blobs and then patch them in other ways (reverse_http(s) used to be this way.
 * Those which implement generate() but do not call super() at all, such as the new reverse_http(s) modules.

With the current implementation of Msf::Payload::Windows.prepends(), modules which call generate() should be calling super(generated_buffer) not the other way around. We either need to mass-review and fix modules which may call super() in the wrong order or find a better way to implement prepends that doesn't require on super() call chain hacks.

I disagree with the use of payload(), because this can lead to some nasty hacks down the road. Let me take a crack at solving this in a way that avoids the use of super() call chains entirely.

**Comment by hmoore-r7** [comment]
BTW: The payload() method will not work when raw blobs are not used in the payload modules at all.

**Comment by hmoore-r7** [comment]
This is turning into a bag of worms. We have over a 100 payloads that define ```generate()``` explicitly, these payloads have no easy way of applying a prepended stub to their output without being modified to return ```apply_prepend(generated_buffer)```. At the same time, the rest of the payloads leverage platform-specific mixins and an overridden call to ```generate()``` in order to apply a prepend to static blob payloads. Any solution that works for one case is not going to work in the other. We can't just redefine ```generate()``` at the top level either, because ```generate()``` is how other classes get the contents of the payload (there isn't a middle-man function where we could apply hooks).

**Comment by hmoore-r7** [comment]
To make things more complicated, it isn't clear where ```Prepend``` and ```PrependEncoder``` options should be applied as it relates to mixin-level prepends. If we need to fix up the heap or align the stack, does this happen before the ```PrependMigrate``` stub, or somewhere else? There are only 5 modules today that depend on PrependMigrate at all right now.

**Comment by hmoore-r7** [comment]
It seems like the least-ugly solution is to pull the prepend logic out of ```generate()``` entirely, change the code which calls ```payload_instance.generate()``` to something like ```payload_instance.generate_complete```, implement ```generate_complete()``` to call ```generate()``` and then pass it through a prepend method if one has been defined. This would also give us a place in the future to do post-processing of generated payload output (dynamic rewriting, that kind of thing).

**Comment by hmoore-r7** [comment]
It looks like we have something very close this already implemented, defined as ```reqs['EncapsulationRoutine']```. This "wraps" the raw payload with a suppled in ```EncodedPayload.generate_raw```. Unfortunately, this wouldn't be usable or convertible, as ```EncapsulationRoutine``` is applied **after** exploit-supplied ```Prepend``` and ```Append``` options, where things like ```PrependMigrate``` would want to be between the ```Prepend``` and ```Append``` options.

**Comment by hmoore-r7** [comment]
I feel like #4918 provides the right solution. Any feedback @scriptjunkie @jlee-r7 @wvu-r7 @todb-r7 @mubix ?

**Comment by scriptjunkie** [comment]
Agreed!
